### PR TITLE
Add DBSCAN example

### PR DESCRIPTION
**Pull request by aprokop**
In #736, the main DBSCAN driver was moved to the `benchmarks/`. It is still a good idea to have an example, so this is its replacement.

### CODE REVIEW

File: examples/dbscan/example_dbscan.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+  // Running with minpts = 5 and eps = 1 would produce no clusters. All the
+  // entries in the labels array would be marked as noise, -1.
+  labels = ArborX::dbscan(ExecutionSpace{}, cloud, 1.f, 5);
+  printLabels(labels);
+
+  // Running with minpts = 3 and eps = 1.5 would produce a single cluster
+  // consisting of points with indices [0, 1, 2, 3, 4, 5, 6, 7].
+  labels = ArborX::dbscan(ExecutionSpace{}, cloud, 1.5f, 2);
+  printLabels(labels);
```
Comment (masterleinad) [inline examples/dbscan/example_dbscan.cpp:73]:
```suggestion
  // Running with minpts = 3 and eps = 1.5 would produce a single cluster
  // consisting of points with indices [0, 1, 2, 3, 4, 5, 6, 7].
  labels = ArborX::dbscan(ExecutionSpace{}, cloud, 1.5f, 3);
  printLabels(labels);
```
?

File: examples/dbscan/example_dbscan.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  // Running with minpts = 4 and eps = 1.5 would produce two clusters
+  // consisting of points with indices [1, 2, 3, 4] and [0, 5, 6, 7]. Point with
+  // index 9 is a border point and may be assigned to the first, or the second
+  // cluster.
```
Comment (masterleinad) [inline examples/dbscan/example_dbscan.cpp:78]:
```suggestion
  // Running with minpts = 4 and eps = 1.5 would produce two clusters
  // consisting of points with indices [1, 2, 3, 4] and [0, 5, 6, 7]. The point with
  // index 9 is a border point and may be assigned to the first, or the second
  // cluster.
```

**Review by masterleinad — COMMENTED** [review]
I think pointing to benchmarks/dbscan/README.md somewhere would be helpful.

File: examples/dbscan/example_dbscan.cpp (hunk 1 above)
Comment (aprokop) [reply examples/dbscan/example_dbscan.cpp:73]: Yeah, misprint. Running with 2 points would also produce a single cluster in this configuration.

**Comment by aprokop** [comment]
> I think pointing to benchmarks/dbscan/README.md somewhere would be helpful.

We have a wiki page for DBSCAN, I think that's sufficient.

**Review by masterleinad — APPROVED** [review]

File: examples/dbscan/example_dbscan.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+using MemorySpace = ExecutionSpace::memory_space;
+
+template <typename MemorySpace>
+void printLabels(Kokkos::View<int *, MemorySpace> labels)
+{
+  auto labels_host =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, labels);
+  for (int i = 0; i < (int)labels_host.size(); ++i)
+    std::cout << labels_host(i) << " ";
+  std::cout << std::endl;
```
Comment (dalg24) [inline examples/dbscan/example_dbscan.cpp:26]: I'd prefer
```suggestion
  std::cout << '\n';
```

File: examples/dbscan/example_dbscan.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  cloud_host[5] = {3.f, 3.f, 0.f};
+  cloud_host[6] = {3.f, 4.f, 0.f};
+  cloud_host[7] = {4.f, 4.f, 0.f};
+  cloud_host[8] = {4.f, 0.f, 0.f};
+  cloud_host[9] = {2.f, 2.f, 0.f};
+  Kokkos::deep_copy(cloud, cloud_host);
+
+  // Running with minpts = 2 and eps = 1 would produce two clusters consisting
+  // of points with indices [1, 2, 3, 4] and [0, 5, 6, 7]. The corresponding
+  // entries in the labels array would be the same.
```
Comment (dalg24) [inline examples/dbscan/example_dbscan.cpp:56]: Should we spell out the output will be `0 1 1 1 1 0 0 0 -1 -1`?

**Review by dalg24 — COMMENTED** [review]
Comment with expected output in the code?

**Comment by aprokop** [comment]
@dalg24 Done.

File: examples/dbscan/example_dbscan.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  // Running with minpts = 5 and eps = 1 would produce no clusters. All the
+  // entries in the labels array would be marked as noise, -1.
+  // Expected output:
+  //   -1 -1 -1 -1 -1 -1 -1 -1 -1 -1
+  labels = ArborX::dbscan(ExecutionSpace{}, cloud, 1.f, 5);
+  printLabels(labels);
+
+  // Running with minpts = 2 and eps = 1.5 would produce a single cluster
+  // consisting of points with indices [0, 1, 2, 3, 4, 5, 6, 7].
```
Comment (dalg24) [inline examples/dbscan/example_dbscan.cpp:70]:
```suggestion
  // consisting of points with indices [0, 1, 2, 3, 4, 5, 6, 7, 9].
```

**Pull request merged**